Clean up CaseAssignmentsController#show_or_hide_contacts and corresponding skipped spec

solebared opened the pull request:
### What github issue is this PR for, if any?
Resolves #5669 and #5670 

### What changed, and _why_?
In addressing issue #5669, it seemed that this endpoint should only be called for in inactive case_assignment. If that premise holds, then we no longer need the `else` conditional that attempts to `render :edit`, which would actually result in an error since this controller doesn't support an edit action or view.

That error was only surfacing in a spec that had been skipped.

If my hypothesis holds, then we can:
* Disallow access to this endpoint via policy if the case_assignment is inactive
* Remove the conditional and lways toggle :hide_old_contacts if authorized
* Change policy and request specs accordingly

### Note
Please verify my hypothesis on this! -- That the ability to show/hide old contacts should not be accessible on an active case_assignment (for any role).

### How is this **tested**? (please write tests!) 💖💪
* This change was motivated by trying to revive a skipped spec, which i believe this PR does.
* The changes to policies also have spec coverage.
* However i haven't reviewed whether we need additional coverage at an integration level.

Steps to test:
* Log in as an admin
* Edit a case with a volunteer assigned
* Unassign the volunteer (bottom of the form)
* Scroll back down to the bottom of the form
* A 'Hide old contacts' button should be visible
* Click it
    * should show a message at the top that contacts are hidden
    * button should now read 'Show old contacts'
    * clicking it again should show another message and toggle state.

### Screenshots please :)
No user facing changes. Here's a screenshot of the bottom of the Case Edit form where the show/hide old contacts page lives:
![Screenshot 2024-06-03 at 12 07 21](https://github.com/rubyforgood/casa/assets/8330/4093b9ea-4b10-497c-9401-c6e5fe7d59b6)

solebared committed:
show_or_hide_contacts endpoint doesn't need to render :edit

In addressing issue #5669, it seemed that this endpoint should only be
called for in inactive case_assignment. If that premise holds, then we
no longer need the `else` conditional that attempts to `render :edit`,
which would actually result in an error since this controller doesn't
support an edit action or view.

That error was surfacing in a spec that had been skipped.

If my hypothesis holds, then we can:
* Disallow access to this endpoint via policy if the case_assignment is
  inactive
* Remove the conditional and always toggle :hide_old_contacts if
  authorized
* Change policy and request specs accordingly

elasticspoon commented:
I think this makes sense to me, I need to look at the code a bit closer but this functionality was added with this PR https://github.com/rubyforgood/casa/pull/3637 to solve this issue https://github.com/rubyforgood/casa/issues/3338. To summarize previously volunteers could see case contacts made by previous volunteers on a case and admins wanted to have the option to disable that.

It would not make sense to toggle this visibility of an active (volunteer) assignment.

elasticspoon commented on app/policies/case_assignment_policy.rb:
```diff
@@ -29,11 +29,11 @@ def reimbursement?
   end
 
   def hide_contacts?
-    !record.active? && !record.hide_old_contacts? && admin_or_supervisor?
+    show_or_hide_contacts? && !record.hide_old_contacts?
   end
```
Something about this really bothers me. This is used in one place. https://github.com/rubyforgood/casa/blob/d7061ea1c9540735038a594422643328e29fb78b/app/views/casa_cases/_volunteer_assignment.html.erb#L59

```rb
<% if (policy(assignment).hide_contacts?) %>
  <%= button_to("Hide Old Contacts",
            show_hide_contacts_case_assignment_path(assignment),
            method: :patch,
            class: "btn btn-outline-secondary text-wrap") %>
<% else %>
  <%= button_to("Show Old Contacts",
            show_hide_contacts_case_assignment_path(assignment),
            method: :patch,
            class: "btn btn-outline-warning text-wrap") %>
<% end %>

```
All it does is dictate whether the show or hide button appears.

But the current logic is incorrect it will show a hide button if the user doesn't have show_hide permissions, which doesn't make sense, instead it should not show a button. We don't actually see this play out because you can't reach that page without the needed permissions.

I feel like this method should not exist and instead the above code should just check `assignement.hide_old_contacts?`

solebared replied:
Fully agree with this. Made it so in 475615fda687a798b6859a48be29d396240e3a2c.

But that got me thinking about other places `hide_old_contacts?` is used...
As you noted, that button toggle was the only place the policy method was used.

But i'm also not seeing any other place that `CaseAssignment#hide_old_contacts?` is used. Does that mean that clicking show/hide old contacts has no actual impact? Where do we actually _hide_ old contacts?

elasticspoon replied:
It looks like the **method** is not used outside of that but the value is used as part of a query.


The show view calls a method off a decorator

https://github.com/rubyforgood/casa/blob/e596c58eb56f8c9b20840b3f5c6ab9aec3ee0daa/app/views/casa_cases/show.html.erb#L135

```rb
<%= render(partial: "case_contacts/case_contact",
          collection: @casa_case.decorate
            .case_contacts_filtered_by_active_assignment_ordered_by_occurred_at
            .includes(contact_types: [:contact_type_group])
            .grab_all(current_user),
          as: :contact) %>
```

That method contains a query that scopes down based on `hide_old_contacts`

https://github.com/rubyforgood/casa/blob/e596c58eb56f8c9b20840b3f5c6ab9aec3ee0daa/app/decorators/casa_case_decorator.rb#L18

```rb
  def case_contacts_filtered_by_active_assignment_ordered_by_occurred_at
    object.case_contacts
      .joins("INNER JOIN case_assignments on case_assignments.casa_case_id = case_contacts.casa_case_id and case_assignments.volunteer_id = case_contacts.creator_id")
      .where("case_assignments.hide_old_contacts = false")
      .order(occurred_at: :desc)
  end
```

elasticspoon reviewed (changes requested):
Overall, I think you are totally correct. I ended up doing some digging and I think we ought to clean up the code a tiny bit since we are there already.

solebared committed:
Remove unnecessary and confusing CaseAssignmentPolicy#hide_contacts?

H/T @elasticspoon:
https://github.com/rubyforgood/casa/pull/5812#discussion_r1625237388

solebared committed:
Add CaseAssignment#inactive? helper

Because assignment.inactive? is easier to grok than !assignment.active?

H/T @elasticspoon:
https://github.com/rubyforgood/casa/pull/5812#discussion_r1625191970

solebared commented on app/policies/case_assignment_policy.rb:
```diff
@@ -29,11 +29,11 @@ def reimbursement?
   end
 
   def show_or_hide_contacts?
-    admin_or_supervisor? && same_org?
```
Added `CaseAssignment.inactive?` in 9165b852aba7954c1279207142caddb36ae8bb60.

solebared committed:
Amend and use same_org method from ApplicationPolicy

solebared commented on app/policies/case_assignment_policy.rb:
```diff
@@ -29,11 +29,11 @@ def reimbursement?
+    !record.active? && admin_or_supervisor? && same_org?
   end
 
   def same_org?
```
Fixed in 279d10870.

solebared commented on app/policies/case_assignment_policy.rb:
Can you say more about what parts of this you're referring to? I've addressed the related comments below; perhaps those fixes address it?

elasticspoon replied:
I was looking for you to move to amend the same_org method in the app policy and use https://github.com/rubyforgood/casa/blob/279d108703248f59d0ba5a722315c2c8ad32ede8/app/policies/application_policy.rb#L103

solebared replied:
I see. Fixed in cc6c95732.

solebared committed:
Use admin_or_supervisor_same_org? from ApplicationPolicy

elasticspoon commented on spec/policies/case_assignment_policy_spec.rb:
```diff
@@ -67,11 +67,12 @@
       end
       context "in a different organization" do
         it "does not allow user to show or hide contacts" do
+          other_case_assignment.update(active: false)
```
This is a good call. Otherwise an admin might not have perms because the case is inactive rather than because the case belongs to another org.
